update

liang-bo96 opened the pull request.

Commit:
refactor

Commit:
refactor

Commit:
update install

Commit:
lint

christianbrodbeck commented on docs/api_reference.rst:
```diff
@@ -18,6 +18,34 @@ Sample Data Module
+
+   register_layout("my_layout", MyLayout())
+
+**LayoutBuilder**
```
Use autodoc

christianbrodbeck commented on docs/user_guide.rst:
```diff
@@ -157,6 +157,23 @@ Horizontal Layout
+Custom Layouts
+^^^^^^^^^^^^^^
+
+Create custom layouts by inheriting from ``LayoutBuilder``:
```
Use sphinx cross-references with auto-doc: https://www.sphinx-doc.org/en/master/usage/domains/python.html#python-xref-roles

christianbrodbeck commented on docs/user_guide.rst:
```diff
@@ -157,6 +157,23 @@ Horizontal Layout
+
+.. code-block:: python
+
+   from eelbrain_plotly_viz import LayoutBuilder, register_layout, EelbrainPlotly2DViz
```
Reduce redundancy. This is already documented under the relevant API reference. Otherwise you need to make sure it's correct in two places.

christianbrodbeck commented on src/eelbrain_plotly_viz/__init__.py:
```diff
@@ -3,6 +3,25 @@
+
+Custom layouts can be created by inheriting from LayoutBuilder:
+
+>>> from eelbrain_plotly_viz import LayoutBuilder, register_layout
```
As above, minimize redundancy

christianbrodbeck commented on src/eelbrain_plotly_viz/__init__.py:
```diff
@@ -11,11 +30,40 @@
     "BrainPlotly2DViz",  # Alias for compatibility
+    # Sample data
     "create_sample_brain_data",
+    # Mixins for extensibility
```
Do these Mixins need to be public? If so, then they should be documented in the API docs.

christianbrodbeck reviewed (changes requested):
A few comments on the docs, but one major question (and I'd be interested in @smiths opinion as well): Why use the mixing pattern? It seems inherently finicky to me due to the lack of proper encapsulation (each mixin can potentially overwrite attributes used by another mixing, unless you use name mangling consistently, which seems clunky too)

liang-bo96 commented:
You’re right that mixins can be problematic, especially around weak encapsulation and the risk of attributes clobbering each other.

In this module I’m mostly using them as an internal structuring tool: there is only one public class, EelbrainPlotly2DViz, and the mixins just split its responsibilities into data loading, figure creation, layout, and app control so the implementation can live in focused files.

The risk of attribute collisions is reduced here mainly in two ways: first, each responsibility has exactly one mixin, so we don’t have multiple mixins competing over the same concern; second, the shared attributes are given descriptive, domain‑specific names and are declared via type hints, so the “shared state contract” is as explicit as possible.

If this still feels too fragile, I’m happy to refactor towards a composition‑based design (separate helper objects managed by EelbrainPlotly2DViz) while keeping the public API unchanged.

Commit:
update

Commit:
remove mixin

Commit:
update

Commit:
use helper

Commit:
use helper

christianbrodbeck commented on docs/api_reference.rst:
```diff
@@ -18,6 +18,19 @@ Sample Data Module
+
+Custom layouts can be created by inheriting from ``LayoutBuilder``:
+
+.. autoclass:: eelbrain_plotly_viz.layout_builder_helper.LayoutBuilder
```
Could this be displayed using the name that users are supposed to use? I think you should be able to write `eelbrain_plotly_viz.VerticalLayout` because you import it in `eelbrain_plotly_viz`; since your namespace is flat `~eelbrain_plotly_viz.VerticalLayout` would be even simpler

christianbrodbeck commented on src/eelbrain_plotly_viz/viz_2d.py:
```diff
@@ -34,84 +31,94 @@ class EelbrainPlotly2DViz:
+    - PlotFactoryHelper: Responsible for constructing all visualization figures
+    - LayoutBuilderHelper: Responsible for arranging UI components
+    - AppControllerHelper: Responsible for user interaction logic
+
```
I would not put this in the docstring because it's not public API. Users don't need to know this, only programmers. You already describe it in the `__init__.py` doctoring, so I would say DRY here.

christianbrodbeck commented on src/eelbrain_plotly_viz/app_controller_helper.py:
```diff
@@ -0,0 +1,395 @@
+    - Exporting visualizations as image files
+    - Running the Dash application server
+
+    The mixin separates interaction logic from data loading, plot creation,
```
"mixin"

christianbrodbeck commented on src/eelbrain_plotly_viz/app_controller_helper.py:
```diff
@@ -0,0 +1,395 @@
+    and layout concerns, ensuring each responsibility is handled independently.
+    """
+
+    # Declare expected attributes from other mixins/main class
```
?

christianbrodbeck reviewed (changes requested):
Looks like the refactor is not complete (see comments), please let me know when it's finished.

Also: can you make all the new modules private, and make sure you only import public names into `__init__.py` what you actually want users to see? `__all__` is not always used, for example all names (not just `__all__`) show up in Jupyter autocomplete, and that 1) clutters the namespace and 2) may cause confusions about what is public and what is private.

Commit:
update

liang-bo96 commented:
I’ve updated code in 67ab16ba @christianbrodbeck

christianbrodbeck commented on src/eelbrain_plotly_viz/__init__.py:
```diff
@@ -3,19 +3,45 @@
 __author__ = "LiveNeuron Team"
 
+# Main visualization class
 from .viz_2d import EelbrainPlotly2DViz
```
`viz_2d` intentionally public?

christianbrodbeck commented on src/eelbrain_plotly_viz/_app_controller_helper.py:
```diff
@@ -0,0 +1,392 @@
+        if name == "_viz":
+            super().__setattr__(name, value)
+        else:
+            setattr(self._viz, name, value)
```
I don't get this. Why set and get all attributes in the `EelbrainPlotly2DViz` instance? Doesn't that contradict the goal of encapsulation? IMO it's also confusing for people trying to read the code.

christianbrodbeck reviewed (changes requested):
Sorry, another very basic question (see comments)

Commit:
update

Commit:
update

Commit:
update

Commit:
update

christianbrodbeck commented on src/eelbrain_plotly_viz/_data_loader_helper.py:
```diff
@@ -0,0 +1,307 @@
+        if region is not None:
+            try:
+                data_ds["src"] = set_parc(data_ds["src"], region)
+                self._viz.region_of_brain = region
```
I thought I mentioned before that this does not do anything as far as I can tell? If I am mistaken, can you give me an example where it changes the visualization?

christianbrodbeck replied:
This?

liang-bo96 replied:
<img width="622" height="255" alt="image" src="https://github.com/user-attachments/assets/4a6ff42c-4179-4b23-be4f-783a415dadf5" />

christianbrodbeck commented on src/eelbrain_plotly_viz/_data_loader_helper.py:
```diff
@@ -0,0 +1,307 @@
+            self._viz.parcellation = None
+
+        # Compute norm for butterfly plot
+        self._viz.butterfly_data = np.linalg.norm(self._viz.glass_brain_data, axis=1)
```
This is still setting parent attributes from inside the helper class, this still violates encapsulation and makes it harder to understand what's happening in `EelbrainPlotly2DViz`. If `EelbrainPlotly2DViz` needs information generated in one of the helper classes it should explicitly retrieve it, it should not magically appear (from the point of view of `EelbrainPlotly2DViz`). 

I think a clean way to handle this would be a model or document class that holds the data, which can then be accessed from wherever it's needed.

Commit:
update

Commit:
update

liang-bo96 commented:
updated in [e460dc3](https://github.com/liang-bo96/LiveNeuron/pull/30/commits/e460dc390eff41abb247fadf23add94ff89f048c) @christianbrodbeck

Commit:
update naming

Commit:
update naming

Commit:
update naming

Commit:
update naming

liang-bo96 commented:
I've fixed the naming issue in [c6609a6](https://github.com/liang-bo96/LiveNeuron/pull/30/commits/c6609a69af81c7db5bf7e122c9fb88b710fac97e) and updated the doc @christianbrodbeck

christianbrodbeck commented on src/liveneuro/_liveneuro.py:
What about renaming this file to `_liveneuro` for consistency?

christianbrodbeck commented on src/liveneuro/_viz_2d.py:
```diff
@@ -0,0 +1,345 @@
+        https://plotly.com/python/builtin-colorscales/ for all available options.
+    vmin
+        Optional lower bound for the color range. If provided, locks the minimum
+        for all projections and time points. Always 0.
```
What does "Always 0" mean? Default is 0? Or it's 0 regardless of what value you supply?

christianbrodbeck commented on src/liveneuro/_viz_2d.py:
```diff
@@ -0,0 +1,345 @@
+        Layout arrangement mode for the visualization interface. Options:
+        - 'vertical': Traditional layout with butterfly plot on top, brain views below (default)
+        - 'horizontal': Compact layout with butterfly plot on left, brain views on right
+        Default is 'vertical' for backward compatibility.
```
Without a significant user base yet you should not worry about backwards compatibility yet. If horizontal makes more sense as default it should be horizontal.

christianbrodbeck commented on src/liveneuro/_data_loader_helper.py:
```diff
@@ -0,0 +1,299 @@
+            parcellation=parcellation,
+        )
+
+    def _parse_display_mode(self, mode: str) -> List[str]:
```
This does not seem to be related to "data loading" but to visualization. It's also a static method. I would turn this into a function and move it into the part of your application concerned with interpreting the basic user inputs or visualization. 

This also applies to all methods below this one

christianbrodbeck commented on src/liveneuro/_layout_helper.py:
```diff
@@ -0,0 +1,821 @@
+    def build(self, app: "LiveNeuro") -> Dict[str, Any]:
+        """Build layout configuration and Dash layout components.
+
+        :param app: LiveNeuro instance to configure layout for.
```
Why are you suddenly using a different documentation style?

You are on a very tight timeline, please don't waste more review cycles with basic things like that.

christianbrodbeck reviewed (changes requested):
Thanks for addressing #31 – it's not consistent in `/INSTALL.md`, but that file can probably be deleted? – for the future, please consider the advice from your contributors guide and keep PRs manageable by not lumping multiple issues into a single PR.

Commit:
update naming

liang-bo96 commented:
Sorry about the earlier issues —I’ve had a bad cold and a persistent fever in recent days and made some careless mistakes.
I’ve now cleaned things up based on your feedback: removed redundant files, renamed files, fixed docstring, and reorganized some methods in [a9abc05](https://github.com/liang-bo96/LiveNeuro/pull/30/commits/a9abc05f6317d7c15fd6910380310749354b6e06).

I also updated the report based on the newest code ,two main commit are https://github.com/liang-bo96/projectReport/commit/db6d4d3e5bb3319e6b79aa18df58fd24d1593212
https://github.com/liang-bo96/projectReport/commit/cc2b388c63b81b4ad93fbd3549e22edc3776524c

Thanks a lot for your patience and help! @christianbrodbeck

christianbrodbeck commented on src/liveneuro/_data_loader_helper.py:
```diff
@@ -0,0 +1,122 @@
+    parcellation: Optional[Any]
+
+
+class DataLoaderHelper:
```
Why is this a class? The only class members are two static methods. Shouldn't these just be functions?

liang-bo96 replied:
DataLoaderHelper could technically be a couple of module-level functions, since it’s currently stateless. The reason I kept it as a class is to match the project’s “helper-by-composition” architecture: data loading is treated as a component with a stable interface related to data. This also leaves a clean extension for future needs.

christianbrodbeck commented on src/liveneuro/_layout_helper.py:
```diff
@@ -0,0 +1,924 @@
+This module provides the LayoutBuilder interface, concrete layout strategies,
+and the LayoutBuilderHelper following the Open/Closed Principle.
+
+New layout strategies can be added by:
+1. Inheriting from LayoutBuilder and implementing the build() method
+2. Registering the new layout with register_layout()
+
+Example
+-------
+>>> from liveneuro import LayoutBuilder, register_layout
+>>>
+>>> class GridLayout(LayoutBuilder):
+...     def build(self, app):
+...         # Custom layout implementation
+...         pass
+>>>
+>>> register_layout("grid", GridLayout())
+>>> viz = LiveNeuro(layout_mode="grid")
```
This seems to be partially duplicated in the [API docs](https://liveneuron--30.org.readthedocs.build/en/30/api_reference.html#liveneuro.LayoutBuilder), but the API docs look weird and the register step is missing from the example

christianbrodbeck replied:
I think you should have at least one working example of this process to make it in the general users' API docs. You could point to your `LayoutBuilder` subclasses, but those look way more complex than "just override `.build`. Given the timeline I suggest, if you don't have a working simpler example, you move all the documentation to this module here and keep it private for now.

liang-bo96 replied:
I've moved the doc and set it private

christianbrodbeck commented on src/liveneuro/_plot_factory_helper.py:
```diff
@@ -0,0 +1,1005 @@
+        n_sources, n_times = self._viz.butterfly_data.shape
+
+        # Auto-scale data for visibility
+        data_to_plot = self._viz.butterfly_data.copy()
```
Do I understand correctly that this happens every time the time course plot gets drawn? This is all work that could be done once and then saved in memory. Also, by the way, you're copying the array here without ever mutating it.

liang-bo96 replied:
added a new cache method in [56bb92c](https://github.com/liang-bo96/LiveNeuro/pull/30/commits/56bb92c2aa21e314d20f59117648d0fda3d24561)

christianbrodbeck commented on push_to_github.sh:
This file seems like your personal workflow and not something that belong in the repository(?) given it hard codes `origin https://github.com/liang-bo96/LiveNeuro.git`

liang-bo96 replied:
removed

christianbrodbeck reviewed (changes requested):
Sorry to hear it got you too.

I still don't think segregation of tasks is ideal, but this PR has become too big to work with properly, and given the timeline there is probably not much more you can do. In retrospect I think this refactor is something we should have discussed and planned in one of our meetings. Next time you do something like this I recommend looking into existing patterns like [Model–view–controller](https://en.wikipedia.org/wiki/Model–view–controller).

One more note, you have a lot of static methods that are not marked as such – if they should really be methods and not functions, mark them as `@staticmethod`.

christianbrodbeck commented:
I would say Address my last comments and then we'll merge so that the repo is consistent with the report, and take it from there. 

If you can do one more thing, I would recommend the following: make sure your API reflects encapsulation by never accessing a private attribute of a class. I.e., you should never have something like `self._data_loader._load_data()` (or `data_loader._load_data()`). If you need that, `._load_data()` should become public (`.load_data()` – that does not mean you need to expose it to the user, as it may be a public attribute of a private class)

Commit:
update naming

Commit:
update naming

Commit:
update naming

Commit:
update static method

Commit:
update static method

christianbrodbeck commented on src/liveneuro/_liveneuro.py:
```diff
@@ -0,0 +1,394 @@
+
+    @property
+    def current_layout_config(self) -> Optional[Dict[str, Any]]:
+        return self._current_layout_config
```
Why not just make `self._current_layout_config` public? I would try to keep things as simple as possible, things should have only one name if at all possible, and access should be as direct as possible.

christianbrodbeck commented on src/liveneuro/_liveneuro.py:
```diff
@@ -0,0 +1,394 @@
+    ):
+        return self._plot_factory.create_butterfly_plot(
+            selected_time_idx=selected_time_idx, figure_height=figure_height
+        )
```
Same here, and following ones, why the aliases? Doesn't that contradict the division of responsibility you were trying to achieve? Also, remember `Flat is better than nested.`

christianbrodbeck reviewed (commented):
A question on the design as I started reading (see comments)

Commit:
update
